pkg/networkservice/chains/xconnectns/server.go

```diff
@@ -56,68 +71,197 @@ import (
-			kernel.MECHANISM:    kernel.NewServer(vppConn),
-			vxlan.MECHANISM:     vxlan.NewServer(vppConn, tunnelIP),
-			wireguard.MECHANISM: wireguard.NewServer(vppConn, tunnelIP),
-		}),
-		pinhole.NewServer(vppConn),
-		connect.NewServer(
-			ctx,
-			client.NewClientFactory(
-				client.WithName(name),
-				client.WithAdditionalFunctionality(
-					mechanismtranslation.NewClient(),
-					connectioncontextkernel.NewClient(),
-					stats.NewClient(ctx),
-					up.NewClient(ctx, vppConn),
-					mtu.NewClient(vppConn),
-					tag.NewClient(ctx, vppConn),
-					// mechanisms
-					memif.NewClient(vppConn),
-					kernel.NewClient(vppConn),
-					vxlan.NewClient(vppConn, tunnelIP),
-					wireguard.NewClient(vppConn, tunnelIP),
-					pinhole.NewClient(vppConn),
-					recvfd.NewClient(),
-					sendfd.NewClient()),
-			),
+// NewServer - returns an implementation of the xconnectns network service supporting following mechanisms:
+// * [local VFIO, local kernel with SR-IOV token] -> [remote NOOP]
+// * [local memif, local kernel without SR-IOV token] -> [remote VXLAN, remote wireguard]
+// * [remote VXLAN, remote wireguard] -> [local memif, local kernel]
+// * [remote NOOP] -> [local NOOP]
+func NewServer(ctx context.Context,
+	name string,
+	authzServer networkservice.NetworkServiceServer,
+	tokenGenerator token.GeneratorFunc,
+	vppConn Connection,
+	tunnelIP net.IP,
+	pciPool resourcepool.PCIPool,
+	resourcePool resourcepool.ResourcePool,
+	sriovConfig *config.Config,
+	vfioDir, cgroupBaseDir string,
+	clientURL *url.URL,
+	clientDialOptions ...grpc.DialOption,
+) endpoint.Endpoint {
+	vppChains := createVPPChains(ctx, vppConn, tunnelIP)
+	sriovChains := createSRIOVChains(pciPool, resourcePool, sriovConfig, vfioDir, cgroupBaseDir)
+	noopChains := createNOOPChains()
+
+	return forwarder.NewServer(ctx, tokenGenerator, clientURL,
+		forwarder.WithName(name),
+		forwarder.WithAuthorizeServer(authzServer),
+		forwarder.WithConnectOptions(
 			connect.WithDialOptions(clientDialOptions...),
 		),
+		forwarder.WithBeforeConnectServers(
+			recvfd.NewServer(),
+			sendfd.NewServer(),
+			mechanisms.NewServer(map[string]networkservice.NetworkServiceServer{
+				vfiomech.MECHANISM: sriovChains.beforeConnectServer,
+				kernel.MECHANISM: switchcase.NewServer(
+					&switchcase.ServerCase{
+						Condition: func(_ context.Context, conn *networkservice.Connection) bool {
+							return kernelmech.ToMechanism(conn.GetMechanism()).GetSRIOVTokenID() != ""
+						},
+						Server: sriovChains.beforeConnectServer,
+					},
+					&switchcase.ServerCase{
+						Condition: switchcase.Default,
+						Server:    vppChains.beforeConnectServer,
+					},
+				),
+				memif.MECHANISM:     vppChains.beforeConnectServer,
+				vxlan.MECHANISM:     vppChains.beforeConnectServer,
+				wireguard.MECHANISM: vppChains.beforeConnectServer,
+				noopmech.MECHANISM:  noopChains.beforeConnectServer,
+			}),
+		),
+		forwarder.WithConnectClients(
```

Comment:
Is there a reason we are doing this Switch at the top level instead of simply rolling it under the kernel mechanism (for kernel) and creating a new mechanism/vfio (for vfio)?

Reply:
We have different underlying mechanisms for providing kernel interface in VPP case and in SR-IOV case.
Actually, SR-IOV kernel is mostly like VFIO, not like VPP kernel - because we need to select VF by token, bound kernel driver for the VF (same part) and then configure IP address, routes in kernel way, not in VPP way (kernel-specific part).